### PR TITLE
🌱Add check to disallow hcloud servers without ssh key

### LINKED ISSUE #1402: Disallow HCloud servers without SSH key
/kind feature

**Describe the solution you'd like**
No SSH key should be usually (otherwise please comment) a configuration issue. We can easily catch that and make users aware of this.

This will introduce a breaking change and will affect everyone provisioning HCloud servers without ssh key.

However, I think the drawbacks are not as severe and the advantage will be that no-one can misconfigure it.

We will add a condition to HCloudMachine that there is no ssh key.

The ssh key can be provided by the feature of adding it to the Hetzner secret even on the fly, without changing any templates.

### PR DESCRIPTION
**Pull request by yrs147**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1402 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squash commits
- [ ] include documentation
- [ ] add unit tests

### CODE REVIEW

File: pkg/services/hcloud/server/server.go [hunk 1]
```diff
@@ -435,6 +435,17 @@ func (s *Service) createServer(ctx context.Context) (*hcloud.Server, error) {
 		return nil, errServerCreateNotPossible
 	}
 
+	if len(sshKeySpecs) == 0 {
+		conditions.MarkFalse(
+			s.scope.HCloudMachine,
+			infrav1.ServerCreateSucceededCondition,
+			"NoSSHKeyFound",
```
Comment (janiskemper) [inline pkg/services/hcloud/server/server.go:442]: can you add the reason to the api folder where the other reasons are defined?

File: pkg/services/hcloud/server/server.go [hunk 2]
```diff
@@ -435,6 +435,17 @@ func (s *Service) createServer(ctx context.Context) (*hcloud.Server, error) {
 		return nil, errServerCreateNotPossible
 	}
 
+	if len(sshKeySpecs) == 0 {
```
Comment (janiskemper) [inline pkg/services/hcloud/server/server.go:438]: this check can be done before you fetch the ssh keys from the API

File: pkg/services/hcloud/server/server.go [hunk 3]
```diff
@@ -435,6 +435,17 @@ func (s *Service) createServer(ctx context.Context) (*hcloud.Server, error) {
...
+		conditions.MarkFalse(
+			s.scope.HCloudMachine,
+			infrav1.ServerCreateSucceededCondition,
+			"NoSSHKeyFound",
+			clusterv1.ConditionSeverityError,
+			"No SSH key configured for the server. Please add an SSH key ",
+		)
+		return nil, fmt.Errorf("SSHKeySpec is empty")
```
Comment (janiskemper) [inline pkg/services/hcloud/server/server.go:446]: return errServerCreateNotPossible

**Review by janiskemper — CHANGES_REQUESTED** [review]

File: pkg/services/hcloud/server/server.go (hunk 1 above)
Comment (yrs147) [reply pkg/services/hcloud/server/server.go]: I saw there is already a reason for that `infrav1.SSHKeyNotFoundReason`

**Review by janiskemper — APPROVED** [review]

**Review by janiskemper — COMMENTED** [review]
Sorry, forgot the unit tests. I believe that we can test this. There should be existing tests probably for related cases. Can you check? Should be about that we set the condition when needed

**Pull request merged**